src/linked_hash_map.rs

```diff
@@ -0,0 +1,509 @@
+        })
+    }
+
+    /// Return the maximum number of key-value pairs the map can hold without reallocating.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use collect::LinkedHashMap;
+    /// let mut map: LinkedHashMap<int, &str> = LinkedHashMap::new();
+    /// let capacity = map.capacity();
+    /// ```
+    pub fn capacity(&self) -> uint {
+        self.map.capacity()
+    }
+
+    /// Remove the oldest map entry.
+    ///
+    /// Can be used in implementation of LRU cache.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use collect::LinkedHashMap;
+    /// let mut map: LinkedHashMap<int, int> = LinkedHashMap::new();
+    /// map.insert(1, 10);
+    /// map.insert(2, 20);
+    /// map.remove_lru();
+    /// assert_eq!(map.get(&1), None);
+    /// assert_eq!(map.get(&2), Some(&20));
+    /// ```
+    #[inline]
+    pub fn remove_lru(&mut self) {
+        if self.len() > 0 {
+            let lru = unsafe { (*self.head).prev };
+            self.detach(lru);
+            self.map.remove(&KeyRef{k: unsafe { &(*lru).key }});
+        }
+    }
+
+    #[inline]
+    fn detach(&mut self, node: *mut LinkedHashMapEntry<K, V>) {
+        unsafe {
+            (*(*node).prev).next = (*node).next;
+            (*(*node).next).prev = (*node).prev;
+        }
+    }
+
+    #[inline]
+    fn attach(&mut self, node: *mut LinkedHashMapEntry<K, V>) {
+        unsafe {
+            (*node).next = (*self.head).next;
+            (*node).prev = self.head;
+            (*self.head).next = node;
+            (*(*node).next).prev = node;
+        }
+    }
+
+    /// Return the number of key-value pairs in the map.
+    pub fn len(&self) -> uint { self.map.len() }
+
+    /// Returns whether the map is currently empty.
+    pub fn is_empty(&self) -> bool { self.len() == 0 }
+
+    /// Clear the map of all key-value pairs.
+    pub fn clear(&mut self) {
+        self.map.clear();
+        unsafe {
+            (*self.head).prev = self.head;
+            (*self.head).next = self.head;
+        }
+    }
+
+    /// A double-ended iterator visiting all key-value pairs in order of insertion.
+    /// Iterator element type is `(&'a K, &'a V)`
+    ///
+    /// # Example
+    /// ```rust
+    /// use collect::LinkedHashMap;
+    ///
+    /// let mut map = LinkedHashMap::new();
+    /// map.insert("a", 10);
+    /// map.insert("b", 20);
+    /// map.insert("c", 30);
+    ///
+    /// let mut iter = map.iter();
+    /// assert_eq!(Some((&"a", &10)), iter.next());
+    /// assert_eq!(Some((&"b", &20)), iter.next());
+    /// assert_eq!(Some((&"c", &30)), iter.next());
+    /// assert_eq!(None, iter.next());
+    /// ```
+    pub fn iter(&self) -> Iter<K, V> {
+        Iter {
+            head: unsafe { (*self.head).prev },
+            tail: self.head,
+            remaining: self.len(),
+        }
+    }
+}
+
+impl<K: Hash + Eq, V> Extend<(K, V)> for LinkedHashMap<K, V> {
+    fn extend<T: Iterator<Item=(K, V)>>(&mut self, mut iter: T) {
```

Comment:
It is not safe, because iterator may contain billion pairs with the same key. Resulting map will have the same size as before and capacity of billion.